diff --git a/bfd/config.bfd b/bfd/config.bfd
index 13d678e1..2f7f7174 100644
--- a/bfd/config.bfd
+++ b/bfd/config.bfd
@@ -220,6 +220,20 @@ esac
 case "${targ}" in
 # START OF targmatch.h
 #ifdef BFD64
+
+  i[3-7]86-*-skift*)
+    targ_defvec=i386_elf32_vec
+    targ_selvecs=
+    targ64_selvecs=x86_64_elf64_vec
+    ;;
+#ifdef BFD64
+  x86_64-*-skift*)
+    targ_defvec=x86_64_elf64_vec
+    targ_selvecs=i386_elf32_vec
+    want64=true
+    ;;
+#endif
+
   aarch64-*-darwin*)
     targ_defvec=aarch64_mach_o_vec
     targ_selvecs="arm_mach_o_vec mach_o_le_vec mach_o_be_vec mach_o_fat_vec"
diff --git a/config.sub b/config.sub
index 5b158ac4..f4e60010 100755
--- a/config.sub
+++ b/config.sub
@@ -1346,7 +1346,7 @@ case $os in
 	     | hpux* | unos* | osf* | luna* | dgux* | auroraux* | solaris* \
 	     | sym* | kopensolaris* | plan9* \
 	     | amigaos* | amigados* | msdos* | newsos* | unicos* | aof* \
-	     | aos* | aros* | cloudabi* | sortix* \
+	     | aos* | aros* | cloudabi* | sortix* | skift* \
 	     | nindy* | vxsim* | vxworks* | ebmon* | hms* | mvs* \
 	     | clix* | riscos* | uniplus* | iris* | isc* | rtu* | xenix* \
 	     | knetbsd* | mirbsd* | netbsd* \
diff --git a/gas/configure.tgt b/gas/configure.tgt
index a4828c4c..73fafe3b 100644
--- a/gas/configure.tgt
+++ b/gas/configure.tgt
@@ -121,6 +121,7 @@ esac
 generic_target=${cpu_type}-$vendor-$os
 # Note: This table is alpha-sorted, please try to keep it that way.
 case ${generic_target} in
+  i386-*-skift*)     fmt=elf;;
   aarch64*-*-elf*)			fmt=elf;;
   aarch64*-*-fuchsia*)			fmt=elf;;
   aarch64*-*-linux*)			fmt=elf em=linux
diff --git a/ld/Makefile.am b/ld/Makefile.am
index 0509c2e5..ab9d6dd9 100644
--- a/ld/Makefile.am
+++ b/ld/Makefile.am
@@ -284,6 +284,7 @@ ALL_EMULATION_SOURCES = \
 	eelf32xstormy16.c \
 	eelf32xtensa.c \
 	eelf_i386.c \
+	eelf_i386_skift.c \
 	eelf_i386_be.c \
 	eelf_i386_fbsd.c \
 	eelf_i386_ldso.c \
@@ -455,6 +456,7 @@ ALL_64_EMULATION_SOURCES = \
 	eelf_k1om.c \
 	eelf_k1om_fbsd.c \
 	eelf_x86_64.c \
+	eelf_x86_64_skift.c \
 	eelf_x86_64_cloudabi.c \
 	eelf_x86_64_fbsd.c \
 	eelf_x86_64_nacl.c \
diff --git a/ld/Makefile.in b/ld/Makefile.in
index 9898392a..ffb27087 100644
--- a/ld/Makefile.in
+++ b/ld/Makefile.in
@@ -771,6 +771,7 @@ ALL_EMULATION_SOURCES = \
 	eelf32xstormy16.c \
 	eelf32xtensa.c \
 	eelf_i386.c \
+	eelf_i386_skift.c \
 	eelf_i386_be.c \
 	eelf_i386_fbsd.c \
 	eelf_i386_ldso.c \
@@ -941,6 +942,7 @@ ALL_64_EMULATION_SOURCES = \
 	eelf_k1om.c \
 	eelf_k1om_fbsd.c \
 	eelf_x86_64.c \
+	eelf_x86_64_skift.c \
 	eelf_x86_64_cloudabi.c \
 	eelf_x86_64_fbsd.c \
 	eelf_x86_64_nacl.c \
diff --git a/ld/configure.tgt b/ld/configure.tgt
index c81bc8a7..d58b7321 100644
--- a/ld/configure.tgt
+++ b/ld/configure.tgt
@@ -45,6 +45,15 @@ targ64_extra_libpath=
 # architecture variants should be kept together even if their names
 # break the alpha sorting.
 case "${targ}" in
+i[3-7]86-*-skift*)
+			targ_emul=elf_i386_skift
+			targ_extra_emuls=elf_i386
+			targ64_extra_emuls="elf_x86_64_skift elf_x86_64"
+			;;
+x86_64-*-skift*)
+			targ_emul=elf_x86_64_skift
+			targ_extra_emuls="elf_i386_skift elf_x86_64 elf_i386"
+			;;
 aarch64_be-*-elf)	targ_emul=aarch64elfb
 			targ_extra_emuls="aarch64elf aarch64elf32 aarch64elf32b armelfb armelf" ;;
 aarch64-*-elf | aarch64-*-rtems*)
@@ -782,6 +791,10 @@ esac
 NATIVE_LIB_DIRS='/usr/local/lib /lib /usr/lib'
 case "${target}" in
 
+*-*-skift*)
+  NATIVE_LIB_DIRS='/System/Libraries'
+  ;;
+
 *-*-dragonfly*)
   NATIVE_LIB_DIRS='/lib /usr/lib /usr/pkg/lib /usr/local/lib'
   ;;
diff --git a/ld/emulparams/elf_i386_skift.sh b/ld/emulparams/elf_i386_skift.sh
new file mode 100644
index 00000000..1369b907
--- /dev/null
+++ b/ld/emulparams/elf_i386_skift.sh
@@ -0,0 +1,4 @@
+. ${srcdir}/emulparams/elf_i386.sh
+GENERATE_SHLIB_SCRIPT=yes
+GENERATE_PIE_SCRIPT=yes
+TEXT_START_ADDR=0x40000000
diff --git a/ld/emulparams/elf_x86_64_skift.sh b/ld/emulparams/elf_x86_64_skift.sh
new file mode 100644
index 00000000..f4097a25
--- /dev/null
+++ b/ld/emulparams/elf_x86_64_skift.sh
@@ -0,0 +1,2 @@
+. ${srcdir}/emulparams/elf_x86_64.sh
+TEXT_START_ADDR=0x40000000
